python/mpt/src/mpt/__init__.py

```diff
@@ -0,0 +1,1553 @@
+            if nibble_path == node.rest_of_key:
+                logger.debug("Path matches leaf key, deleting leaf")
+                return None, True
+
+            raise ValueError(
+                f"Delete - Path does not match leaf key: {nibble_path_to_hex(nibble_path)} != {nibble_path_to_hex(node.rest_of_key)}"
+            )
+
+        raise ValueError(f"Unknown node type: {type(node)}")
+
+    def upsert(
+        self, path: Bytes, value: Bytes, root_hash: Optional[Hash32] = None
+    ) -> Hash32:
+        """
+        Insert or update a value in the trie at the given path.
+
+        The upsert process in a Merkle Patricia Trie follows these steps:
+
+        1. Special case - Empty trie:
+        - If the trie is empty (root_hash == EMPTY_TRIE_ROOT_HASH)
+        - Create a new leaf node with the entire path and value
+        - No need for complex path splitting or node creation
+
+        2. Path traversal and node creation:
+        a) At a Branch node:
+            - If path is empty: (not supported in this implementation)
+            - Otherwise: Follow the next nibble
+            - If that branch is empty: Create new leaf node
+            - If branch exists: Recursively upsert into that branch
+
+        b) At an Extension node:
+            - Find common prefix between path and node's key_segment
+            - If paths diverge: Create a branch node at divergence point
+            - If extension is prefix of path: Recursively upsert remaining path
+            - Handle path compression by merging extension nodes when possible
+
+        c) At a Leaf node:
+            - If paths match exactly: Update value
+            - If paths differ: Create branch node at first different nibble
+            - Add both paths (existing and new) to the branch
+            - Create extension node if common prefix exists
+
+        3. Node encoding and storage:
+        - Encode modified nodes using RLP
+        - For nodes ≥ 32 bytes: Store separately and use hash as reference
+        - For nodes < 32 bytes: Store directly in parent node
+        - Update node hashes up to root
+
+        Example of inserting 'abc':123 into a trie:
+        ```
+        Before:                After:
+        [Root]                [Root]
+        |                     |
+        [Branch]              [Branch]
+        |                   /    \
+        'd'               'ab'    'd'
+        |                 |      |
+        456              'c':123  456
+        ```
+
+        Parameters
+        ----------
+        path : Bytes
+            The path where the value should be stored
+        value : Bytes
+            The RLP-encoded value to store
+        root_hash : Hash32, optional
+            The root hash to start from, defaults to the trie's state_root
+
+        Returns
+        -------
+        Hash32
+            The new root hash after insertion/update
+        """
+        if root_hash is None:
+            root_hash = self.state_root
```

Comment:
why this?

Reply:
some kind of fn overload where if the root_hash isn't provided as argument in the function, then it's a state access